crates/nu-command/src/network/post.rs

```diff
@@ -0,0 +1,338 @@
+                description: "Post content to url.com, with username and password",
+                example: "post -u myuser -p mypass url.com 'body'",
+                result: None,
+            },
+        ]
+    }
+}
+
+struct Arguments {
+    path: Option<Value>,
+    body: Option<Value>,
+    raw: bool,
+    insecure: Option<bool>,
+    user: Option<String>,
+    password: Option<String>,
+    content_type: Option<String>,
+    content_length: Option<String>,
+}
+fn run_post(
+    engine_state: &EngineState,
+    stack: &mut Stack,
+    call: &Call,
+    _input: PipelineData,
+) -> Result<nu_protocol::PipelineData, nu_protocol::ShellError> {
+    let args = Arguments {
+        path: Some(call.req(engine_state, stack, 0)?),
+        body: Some(call.req(engine_state, stack, 1)?),
+        raw: call.has_flag("raw"),
+        user: call.get_flag(engine_state, stack, "user")?,
+        password: call.get_flag(engine_state, stack, "password")?,
+        insecure: call.get_flag(engine_state, stack, "insecure")?,
+        content_type: call.get_flag(engine_state, stack, "content_type")?,
+        content_length: call.get_flag(engine_state, stack, "content_length")?,
+    };
+    helper(engine_state, stack, call, args)
+}
+// Helper function that actually goes to retrieve the resource from the url given
+// The Option<String> return a possible file extension which can be used in AutoConvert commands
+fn helper(
+    engine_state: &EngineState,
+    stack: &mut Stack,
+    call: &Call,
+    args: Arguments,
+) -> std::result::Result<PipelineData, ShellError> {
+    let url_value = if let Some(val) = args.path {
+        val
+    } else {
+        return Err(ShellError::UnsupportedInput(
+            "Expecting a URL as a string but got nothing".to_string(),
+            call.head,
+        ));
+    };
+    let body = if let Some(body) = args.body {
+        body
+    } else {
+        return Err(ShellError::UnsupportedInput(
+            "Expecting a body parameter but got nothing".to_string(),
+            call.head,
+        ));
+    };
+    let span = url_value.span()?;
+    let requested_url = url_value.as_string()?;
+    let url = match url::Url::parse(&requested_url) {
+        Ok(u) => u,
+        Err(_e) => {
+            return Err(ShellError::UnsupportedInput(
+                "Incomplete or incorrect URL. Expected a full URL, e.g., https://www.example.com"
+                    .to_string(),
+                span,
+            ));
+        }
+    };
+    let user = args.user.clone();
+    let password = args.password;
+    let location = url;
+    let raw = args.raw;
+    let login = match (user, password) {
+        (Some(user), Some(password)) => Some(encode(&format!("{}:{}", user, password))),
+        (Some(user), _) => Some(encode(&format!("{}:", user))),
+        _ => None,
+    };
+    let mut request = http_client(args.insecure.is_some()).post(location);
+    match body {
+        Value::Binary { val, .. } => {
```

Comment:
NOTE: I repeat myself a few times here. Perhaps a macro could avoid repetition in the future.
request.body is quite flexible in what it can accept.